packages/core/src/MachineNode.ts

```diff
@@ -0,0 +1,351 @@
+   *
+   * @param state The state to resolve
+   */
+  public resolveState(state: State<TContext, TEvent>): State<TContext, TEvent> {
+    const configuration = Array.from(
+      getConfiguration([], getStateNodes(this, state.value))
+    );
+    return new State({
+      ...state,
+      value: resolveStateValue(this, state.value),
+      configuration
+    });
+  }
+
+  /**
+   * Determines the next state given the current `state` and sent `event`.
+   *
+   * @param state The current State instance or state value
+   * @param event The event that was sent at the current state
+   * @param context The current context (extended state) of the current state
+   */
+  public transition(
+    state: StateValue | State<TContext, TEvent> = this.initialState,
+    event: Event<TEvent> | SCXML.Event<TEvent>
+  ): State<TContext, TEvent, TStateSchema, TTypestate> {
+    const _event = toSCXMLEvent(event);
+    let currentState: State<TContext, TEvent>;
+
+    if (state instanceof State) {
+      currentState = state;
+    } else {
+      const resolvedStateValue = resolveStateValue(this, state);
+      const resolvedContext = this.machine.context!;
+
+      currentState = this.resolveState(
+        State.from<TContext, TEvent>(resolvedStateValue, resolvedContext)
+      );
+    }
+
+    if (!IS_PRODUCTION && _event.name === WILDCARD) {
+      throw new Error(`An event cannot have the wildcard type ('${WILDCARD}')`);
+    }
+
+    if (this.strict) {
+      if (!this.events.includes(_event.name) && !isBuiltInEvent(_event.name)) {
+        throw new Error(
+          `Machine '${this.id}' does not accept event '${_event.name}'`
+        );
+      }
+    }
+
+    const stateTransition = transitionNode(
+      this,
+      currentState.value,
+      currentState,
+      _event
+    ) || {
+      transitions: [],
+      configuration: [],
+      entrySet: [],
+      exitSet: [],
+      source: currentState,
+      actions: []
+    };
+
+    const prevConfig = getConfiguration(
+      [],
+      getStateNodes(this, currentState.value)
+    );
+    const resolvedConfig = stateTransition.configuration.length
+      ? getConfiguration(prevConfig, stateTransition.configuration)
+      : prevConfig;
+
+    stateTransition.configuration = [...resolvedConfig];
+
+    return resolveTransition(this, stateTransition, currentState, _event);
+  }
+
+  /**
+   * The initial State instance, which includes all actions to be executed from
+   * entering the initial state.
+   */
+  public get initialState(): State<TContext, TEvent, TStateSchema, TTypestate> {
+    this._init();
+    const { initialStateValue } = this;
+
+    if (!initialStateValue) {
+      throw new Error(
+        `Cannot retrieve initial state from simple state '${this.id}'.`
+      );
+    }
+
+    return getInitialState(this, initialStateValue);
+  }
+
+  public getStateNodeById(id: string): StateNode<TContext, any, TEvent> {
```

Comment:
Many methods have been removed in favor of state utils - should we do the same with this one?

Reply:
Probably but not quite yet, since this one is used frequently. I want to figure out the import story first (flat imports, `xstate/utils` instead of `xstate/lib/utils`)